File: fixtures/java/Ledger.java
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Ledger {
    public static final int RETRY_LIMIT = 3;

    private static String region = "us-east-1";

    private final Map<String, Double> cache = new HashMap<>();

    public static class Entry {
        public String id;
        public double amount;
        public String kind;
    }

    public static class Store {
        protected final Map<String, Double> entries = new HashMap<>();

        public void put(String id, double amount) {
            entries.put(id, amount);
        }

        public double total(String id) {
            return entries.get(id);
        }
    }

    public static class CappedStore extends Store {
        @Override
        public void put(String id, double amount) {
            if (entries.size() >= 2) {
                return;
            }
            entries.put(id, amount);
        }
    }

    public boolean validate(Entry entry) {
        if (entry.id == null) {
            return false;
        }
        return entry.amount >= 0;
    }

    public String handle(Entry entry) {
        if (entry.kind.equals("created")) {
            return onCreated(entry);
        } else if (entry.kind.equals("updated")) {
            return onUpdated(entry);
        } else {
            return onCreated(entry);
        }
    }

    private String onCreated(Entry entry) {
        return "ok";
    }

    private String onUpdated(Entry entry) {
        return "ok";
    }

    public Entry persist(Entry entry) {
        if (entry.id == null) {
            throw new IllegalArgumentException("invalid record");
        }
        cache.put(entry.id, entry.amount);
        return entry;
    }

    public Entry process(Entry entry) {
        if (!validate(entry)) {
            System.out.println("bad record");
            return null;
        }
        try {
            return persist(entry);
        } catch (Exception e) {
        }
        return null;
    }

    public List<String> ids() {
        return new ArrayList<>(cache.keySet());
    }

    public String stamp(String id) {
        return id + "@" + Instant.now().toEpochMilli();
    }

    public static void setRegion(String r) {
        region = r;
    }

    public static String region() {
        return region;
    }

    public String renderRow(String label, double amount, String unit,
                            int precision, String align, int width, char fill) {
        String text = String.format("%." + precision + "f%s", amount, unit);
        String body = label + text;
        StringBuilder pad = new StringBuilder();
        for (int i = body.length(); i < width; i++) {
            pad.append(fill);
        }
        if (align.equals("right")) {
            return pad + body;
        }
        return body + pad;
    }
}

File: fixtures/kotlin/Ledger.kt
import java.time.Instant

const val RETRY_LIMIT = 3

var region = "us-east-1"

sealed class Entry {
    data class Created(val id: String, val amount: Double) : Entry()
    data class Updated(val id: String, val amount: Double) : Entry()
    data class Deleted(val id: String) : Entry()
}

open class Store {
    protected val entries = mutableMapOf<String, Double>()

    open fun put(id: String, amount: Double) {
        entries[id] = amount
    }

    fun total(id: String): Double = entries.getValue(id)

    fun ids(): List<String> = entries.keys.toList()
}

class CappedStore : Store() {
    override fun put(id: String, amount: Double) {
        if (entries.size >= 2) return
        entries[id] = amount
    }
}

class Ledger {
    private val store = Store()
    private val cache = mutableMapOf<String, Double>()

    fun apply(entry: Entry): String = when (entry) {
        is Entry.Created -> {
            store.put(entry.id, entry.amount)
            "ok"
        }
        is Entry.Updated -> {
            store.put(entry.id, entry.amount)
            "ok"
        }
        is Entry.Deleted -> "ok"
    }

    fun handle(kind: String): String {
        return if (kind == "created") {
            onCreated()
        } else if (kind == "updated") {
            onUpdated()
        } else {
            onCreated()
        }
    }

    private fun onCreated(): String = "ok"

    private fun onUpdated(): String = "ok"

    fun validate(id: String?, amount: Double): Boolean {
        if (id == null) return false
        return amount >= 0
    }

    fun persist(id: String, amount: Double): Double {
        require(id.isNotEmpty()) { "invalid record" }
        cache[id] = amount
        return amount
    }

    fun process(id: String?, amount: Double): Double? {
        if (!validate(id, amount)) {
            println("bad record")
            return null
        }
        return try {
            persist(id!!, amount)
        } catch (e: Exception) {
            null
        }
    }

    fun stamp(id: String): String = "$id@${Instant.now().toEpochMilli()}"

    fun fetchDeadline(): Long = Instant.now().toEpochMilli() + 30_000

    fun renderRow(
        label: String, amount: Double, unit: String, precision: Int,
        align: String, width: Int, fill: Char
    ): String {
        val text = "%.${precision}f%s".format(amount, unit)
        val body = label + text
        return if (align == "right") body.padStart(width, fill) else body.padEnd(width, fill)
    }
}
